### ISSUE TITLE
Mods can easily bypass security if there is a trusted mod

### ISSUE THREAD

**Issue by est31** [minetest/minetest#3132]
Steps to reproduce:
1. create a trusted mod A, which uses some minetest.something function during loading
2. create an untrusted mod B, which gets loaded before A
3. let B override minetest.something with sth like:
```
minetest.something = function (whatever)
    local root = minetest.request_insecure_environment()
    root.do_evil_thing()
end
```
Now B can do what it wants with the environment it has got.

Found by forum user `ArguablySane`: https://forum.minetest.net/viewtopic.php?p=189354#p189354

**Comment by nerzhul** [comment]
Exact

**Comment by est31** [comment]
A fix for this would be to check that every mod's code inside the whole stack is trusted. But the question is how to get the stack, without using the "unstable" `debug` table.

In the longer term its better to have our own `minetest` environment for trusted mods.

**Comment by PilzAdam** [comment]
Maybe pass the insecure environment as an argument to init.lua instead of using `minetest.request_insecure_environment()`?

**Comment by est31** [comment]
@PilzAdam : That might work. We should however help mods with how to store and use that environment.

**Comment by t4im** [comment]
Wouldn't it be cleaner to run every mod in its own environment to begin with?
While reducing the surface for intermod modifications, it would reduce the attack surface as well.

The malicious code would define the request automatically against its own env in the scenario described, despite the unrelated caller.

I'm sure this "every mod in its own context" has come up before, but that's probably because component development reacts very well to visibility hiding.

Intermod communication could be better mediated, which might enforce some better practices and opens up a few new possibilities. The status quo and lua's flexibility might even allow for a lot of backward compatibility here.

**Comment by est31** [comment]
@t4im that would be the best solution, if you find out how to do it so that nothing breaks :)

**Comment by t4im** [comment]
Well, what do you see breaking?

* If you have everything in its own environment with (optionally secured) `_G`-index, you still are able to access everything that you are already able to access.
* A custom `__newindex` still can allow you to expose anything to the shared env. (We gain e.g. the option to limit it to the first table though, if we wish)

`__newindex` will not easily allow you to replace globals of others — but why would you want to do that? (Unless you're [shakedown](https://github.com/t4im/shakedown), for which I'm sure we can get some extra treatment going. Or I'll hate myself for even proposing all this. Heh); However you can still replace or extend functions within the accessible tables "string" or "minetest" (for e.g. `is_protected()`).

Additional to the trusted-only request call, you now gained the constraint of it being called within the environment of the same trusted mod it was provided for:
Requiring the malicious mod not just to trick the trusted mod to execute the request, but to execute it in the _right_ trusted environment.

But that still leaves functions, where we effectively pass the environment as parameter.

Mod specific environments should be able to catch these cases by denying request of an insecure environment after any `newindex` or `index` was called. (Maybe after whitelisting a few things, that are sure not to expose anything on their own, like dofile).

**Comment by dsohler** [comment]
… or simply remove `request_insecure_environment()` without replacement and don’t allow ANY mod to do “evil things” regardless how trustworthy they claim to be?

**Comment by ShadowNinja** [comment]
:'(  Security's hard...

The argument option sounds like a good solution.
Splitting up mods into their own environments would be better (and this has been discussed before, as it could help for a number of things, eg, mod threading), but that would probably have performance implications (and of course it would break all but the most basic mods, due to the large amount of cross-mod API usage, which would still have to be somehow possible, perhaps through something like `minetest.require("mods/WorldEdit")`).

**Comment by rubenwardy** [comment]
Giving each mod its own environment reduces the ability to override other mods, what if you want to override minetest.register_node to add groups to nodes after your mod loads, for example to make all nodes undiggable.

> or simply remove request_insecure_environment() without replacement and don’t allow ANY mod to do “evil things” regardless how trustworthy they claim to be?

It would break:

* IRC mods
* Mods for communicating with server scripts via sockets
* Online stores for eg schematics

If you don't want to give a mod an insecure environment, then don't.

**Comment by t4im** [comment]
> what if you want to override minetest.register_node

Custom environments would not reduce that ability for at all, as long as they make `minetest` available, which they have to anyway.
Just like already mentioned:

> However you can still replace or extend functions within the accessible tables "string" or "minetest" (for e.g. is_protected()).

**Comment by t4im** [comment]
> and of course it would break all but the most basic mods, due to the large amount of cross-mod API usage, which would still have to be somehow possible

I'm not sure that is actually an issue. You can have custom environments and not limit access to `_G` yet (`__index`). Of course that would be even better, if we could limit access, but it is not necessary to already have some of the benefits of a custom environment, including better security.

**Comment by rubenwardy** [comment]
How does this fix the problem, if you can still override minetest.*? Are you talking about removing request_insecure_environment, and just setting an environment for a mod as insecure anyway when it's in the config?

**Comment by t4im** [comment]
`request_insecure_environment` would not be a function in the `minetest`-table anymore but in the environment-table, which can only be called if you actually have access to that environment.

**Comment by ShadowNinja** [comment]
@t4im: As far as I know you can't access one table from multiple `lua_State`s.

**Comment by t4im** [comment]
Ah, i see you were thinking already a bit further. I was only talking about giving them with `lua_setfenv` their own function environments for now.

**Comment by ShadowNinja** [comment]
@t4im: Ah, that might have other security issues though...

5.2 didn't remove function environments, it just modified the syntax and semantics a bit.

**Comment by est31** [comment]
Okay, what about the following: trusted mods get have to provide a `trusted_init.lua`, which gets run in their very own LUA VM context. The `init.lua` file gets run in the normal context again. Between the trusted context and the normal context there is no sharing of globals or other stuff, except one method to pass over strings.

Then creators of trusted mods think twice before they run something in the shared and possibly compromised environment, because they can do everything in the trusted environment already.

Also, they can't just easily store the trusted environment in a global variable in the shared context, because there they never get access.

**Comment by dsohler** [comment]
Btw. who defines what’s a “trusted mod”? An evil mod maker could just claim that the mod is trustworthy and it is automatically allowed for the mod to do evil things?

Before we waste our precious time on technical details we should first make clear what a “trusted mod” is and who decides if it’s trusted or not.

**Comment by rubenwardy** [comment]
The user decides by adding the mod name to secure.trusted_mods in minetest.conf

**Comment by iqualfragile** [comment]
idea: move the untrusted functionality out of lua, as sandboxing is hard to do and easy to break.
Allow every mod whatsoever to communicate with a local socket, and have some program listening on the other side, doing the work.

**Comment by ShadowNinja** [comment]
@iqualfragile: It's already sandboxed from C++.  I hon't know what you mean be giving every mod a socket.  How would that help?

**Comment by est31** [comment]
> I don't know what you mean be giving every mod a socket. How would that help?

I guess @iqualfragile talks about posix sockets. Even windows has support for winsock.

In fact the idea would be really cool, it would allow mods to expose "safe only" interfaces to the lua part of the mod, e.g. the irc mod only passes over the chat messages and an interface to the minetest config. Both are pretty harmless and even if an attacker "took over" the irc mod, they wouldn't do much harm.

It does make writing "safe only" mods harder though, dunno whether that's good (less mods escaping the sandbox) or whether its bad (less mods with "cool" features).

**Comment by paramat** [comment]
#3738 merged 4827ee1258ac9d68808ca4e2a9cb88bf49473e6b

**Comment by numberZero** [comment]
Writable globals can still be exploited. E.g. the [IRC mod](https://github.com/minetest-mods/irc) can be fooled using the following code *in any mod* that happens to load before IRC:
```lua
local real_loadfile = loadfile
local stub_require = require

function loadfile(name)
        local result = real_loadfile(name)
        if type(result) ~= "function" then
                return result
        end
        return function(ie, ...)
                if type(ie) == "table" and ie.require and ie.require ~= stub_require then
                        print("Caught insecure environment!", dump(ie))
                        print("YOU'RE HACKED!")
                end
                return result(ie, ...)
        end
end
```
Metatables might be exploited as well, although that’s more complicated.
